src/viam/sessions_client.py

```diff
@@ -0,0 +1,131 @@
+import asyncio
+from datetime import timedelta
+from typing import Optional
+
+from grpclib import Status
+from grpclib.client import Channel
+from grpclib.events import RecvTrailingMetadata, SendRequest, listen
+from grpclib.exceptions import GRPCError, StreamTerminatedError
+from grpclib.metadata import _MetadataLike
+
+from viam import _TASK_PREFIX, logging
+from viam.proto.robot import RobotServiceStub, SendSessionHeartbeatRequest, StartSessionRequest, StartSessionResponse
+
+LOGGER = logging.getLogger(__name__)
+SESSION_METADATA_KEY = "viam-sid"
+
+
+async def delay(coro, seconds):
+    await asyncio.sleep(seconds)
+    await coro
+
+
+class SessionsClient:
+    """
+    A Session allows a client to express that it is actively connected and
+    supports stopping actuating components when it's not.
+    """
+
+    _current_id: str = ""
+    _disabled: bool = False
+    _lock = asyncio.Lock()
+    _supported: Optional[bool] = None
+    _heartbeat_interval: Optional[timedelta] = None
+
+    def __init__(self, channel: Channel, *, disabled: bool = False):
+        self.channel = channel
+        self.client = RobotServiceStub(channel)
+        self._disabled = disabled
+
+        listen(self.channel, SendRequest, self._send_request)
+        listen(self.channel, RecvTrailingMetadata, self._recv_trailers)
+
+    def reset(self):
+        if self._lock.locked():
+            return
+
+        LOGGER.debug("resetting session")
+        self._supported = None
+
+    async def _send_request(self, event: SendRequest):
+        if self._disabled:
+            return
+
+        if event.method_name in [self.client.StartSession.name]:
```

Comment:
do we need to exempt the other methods at https://github.com/viamrobotics/rdk/blob/main/session/doc.go#L71?

Reply:
ah didn't see this before, will add

Reply:
https://github.com/viamrobotics/viam-python-sdk/pull/301/commits/9f0c4e0f80406612d83d205c4d6df3e9da985fb2